lib/foreman/http_proxy/rest_client_extension.rb

```diff
@@ -9,6 +9,27 @@ def proxy_uri
         super
       end
     end
+
+    module RestClientRefinements
+      refine RestClient::Request do
+        def net_http_object(hostname, port)
```

Comment:
I should have realized this earlier, but this is a reproducer that broke prior to this:

```ruby
u = URI('http://example.com')
u.user = 'userwith:'
URI(u.to_s).user
```

That returns `userwith` rather than `userwith:`. Escaping here is a good solution.

Rather than patching `net_http_object`, should we patch `proxy_uri` instead? I looked at the code:
https://github.com/rest-client/rest-client/blob/2c72a2e77e2e87d25ff38feba0cf048d51bd5eca/lib/restclient/request.rb#L423-L462

Would it be good to enhance it to accept an URI object instead of a string?
```ruby
def proxy_uri
  if defined?(@proxy)
    if @proxy
      @proxy.is_a?(URI) ? @proxy : URI.parse(@proxy)
    else
      false
    end
  elsif RestClient.proxy_set?
    if RestClient.proxy
      RestClient.proxy.is_a?(URI) ? RestClient.proxy : URI.parse(RestClient.proxy)
    else
      false
    end 
  else
    nil 
  end 
end
```

This would be a safe monkey patch and perhaps even a patch that's accepted upstream. Then we can add a method on `HttpProxy` to return an `URI` instance instead.

It still means all API usage needs to be modified (including in plugins).

Reply:
If I recall correctly URI raises an exception if you try to set a string with special characters as user or password so it wouldn't really help us, even though I'd like it more as an interface instead of passing strings around. Or am I missing something?

Reply:
Yep, URI does not allow special characters in the username field:
```
[2] pry(main)> uu = URI('http://s.b.c')
=> #<URI::HTTP http://s.b.c>
[3] pry(main)> uu.user = '@#$%^&*'
URI::InvalidComponentError: bad component(expected userinfo component or user component): @#$%^&*
from /opt/rh/rh-ruby25/root/usr/share/ruby/uri/generic.rb:404:in `check_user'
[4] pry(main)> uu.user = 'aaaa'
=> "aaaa"
```

Of course you can CGI.escape the names, but then you will need to unescape them in RestClient/Net::HTTP classes.

Reply:
Ah, good point. In that case I'm convinced the current implementation is actually the correct one. Given most usernames and passwords don't contain any encoded bits, I think refinements may be complicated when you want to remain compatible with multiple Foreman releases. I think I'd be fine either way (no refinements, always monkey patch vs refinements + opt-in patching.